plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/ForceApply.hs

```diff
@@ -0,0 +1,186 @@
+{- | The 'ForceApply' optimisation pushes 'Force' inside its direct 'Apply' subterms,
+ removing any 'Delay' at the top of the body of the underlying lambda abstraction.
+ For example, @force ((\x -> delay b) a)@ is transformed into @(\x -> b) a@.
+ In such simple cases, the transformation is obviously correct, the question remains
+ if this approach can be generalised.
+
+ Since UPLC programs are created from erasing the types of TPLC programs (see
+ "PlutusCore.Compiler.Erase") we will consider TPLC terms of the following structure:
+
+ @
+   /\T1 -> \X1 -> /\T2 -> \X2 -> /\T3 -> \X3 -> ... -> /\Tn -> \Xn -> body
+ @
+ where @T1 ... Tn@ are lists of type variables (e.g. @T1@ could be @[t, q, p]@)
+ and @X1 ... Xn@ are lists of term variables. Of course, each @/\@ and @\@ here would
+ desugar to a sequence of type/term abstractions.
+
+ In order to reason about the proposed optimisation we need to consider such terms in
+ the context of them being applied to some sequence of terms.
+
+ One important observation is that this transformation requires that the underlying
+ (term)-lambda abstraction will be fully reduced by the applications.
```

Comment:
Not just fully reduced, it should be exactly reduced, right? I.e. `force ((\x y -> delay b) e)` is as erroneous as `force ((\x -> delay b) e1 e2)`, right? Or am I being silly?

Reply:
Oh that's a good point, yes, "exactly reduced" is what I meant. I'll fix.